internal/provider/provider.go

```diff
@@ -127,102 +126,112 @@ func New(version string, st state.StateI) (func() *schema.Provider, subi.Substra
-	state         state.StateI
+type threefoldPluginClient struct {
+	twinID          uint32
+	mnemonics       string
+	substrateURL    string
+	rmbRedisURL     string
+	useRmbProxy     bool
+	gridProxyClient proxy.Client
+	rmb             rmb.Client
+	substrateConn   subi.SubstrateExt
+	manager         subi.Manager
+	identity        subi.Identity
+	state           state.StateI
 }
 
 func providerConfigure(st state.StateI) (func(ctx context.Context, d *schema.ResourceData) (interface{}, diag.Diagnostics), subi.SubstrateExt) {
 	var substrateConn subi.SubstrateExt
 	return func(ctx context.Context, d *schema.ResourceData) (interface{}, diag.Diagnostics) {
-		rand.Seed(time.Now().UnixNano())
 		var err error
-		apiClient := apiClient{}
-		apiClient.mnemonics = d.Get("mnemonics").(string)
-		key_type := d.Get("key_type").(string)
+		threefoldPluginClient := threefoldPluginClient{}
+		threefoldPluginClient.mnemonics = d.Get("mnemonics").(string)
+		if err := validateMnemonics(threefoldPluginClient.mnemonics); err != nil {
+			return nil, diag.FromErr(err)
+		}
+
+		keyType := d.Get("key_type").(string)
 		var identity subi.Identity
-		if key_type == "ed25519" {
-			identity, err = subi.NewIdentityFromEd25519Phrase(string(apiClient.mnemonics))
-		} else if key_type == "sr25519" {
-			identity, err = subi.NewIdentityFromSr25519Phrase(string(apiClient.mnemonics))
-		} else {
+
+		switch keyType {
+		case "ed25519":
+			identity, err = subi.NewIdentityFromEd25519Phrase(string(threefoldPluginClient.mnemonics))
+		case "sr25519":
+			identity, err = subi.NewIdentityFromSr25519Phrase(string(threefoldPluginClient.mnemonics))
+		default:
 			err = errors.New("key_type must be one of ed25519 and sr25519")
 		}
+
 		if err != nil {
 			return nil, diag.FromErr(errors.Wrap(err, "error getting identity"))
 		}
-		sk, err := identity.KeyPair()
+
+		keyPair, err := identity.KeyPair()
 		if err != nil {
 			return nil, diag.FromErr(errors.Wrap(err, "error getting user secret"))
 		}
-		apiClient.identity = identity
+
+		threefoldPluginClient.identity = identity
 		network := d.Get("network").(string)
 		if network != "dev" && network != "qa" && network != "test" && network != "main" {
 			return nil, diag.Errorf("network must be one of dev, qa, test, and main")
 		}
-		apiClient.substrate_url = SUBSTRATE_URL[network]
-		rmb_proxy_url := RMB_PROXY_URL[network]
-		substrate_url := d.Get("substrate_url").(string)
-		passed_rmb_proxy_url := d.Get("rmb_proxy_url").(string)
-		if substrate_url != "" {
-			log.Printf("substrate url is not null %s", substrate_url)
-			apiClient.substrate_url = substrate_url
+		threefoldPluginClient.substrateURL = SUBSTRATE_URL[network]
+		rmbProxyURL := RMB_PROXY_URL[network]
+		substrateURL := d.Get("substrate_url").(string)
+		passedRmbProxyURL := d.Get("rmb_proxy_url").(string)
+		if len(strings.TrimSpace(substrateURL)) != 0 {
+			log.Printf("substrate url is not null %s", substrateURL)
+			threefoldPluginClient.substrateURL = substrateURL
```

Comment:
I believe we should do more validations on the validity of the urls passed to the plugin